### ISSUE TITLE
Launching a game which executable is in a subdirectory of the main directory

### ISSUE THREAD

**Issue by Aqa-Ib** [Open-Wine-Components/ULWGL#63]
Imagine the following folder structure for a game:
- Game main folder: /home/user/game
- Executable folder: /home/user/game/x64/launcher.exe

In this case, the main game directory is different than the executable directory.

This game won't work when launched like this:
- `cd /home/user/game/`
- `ulwgl-run /home/user/game/x64/launcher.exe`

Using vanilla wine, a solution is to run it like this: 
- `cd /home/user/game/x64/`
- `wine start /d "/home/user/game" launcher.exe`

Using Steam, we have the "Start in" field available:

![image](https://github.com/Open-Wine-Components/ULWGL-launcher/assets/16420574/fe32682c-525e-444f-9c3d-6df599d272c4)

Using Lutris we have a "Working directory" field.

How do we achieve this using `ulwgl-run` ?

**Comment by Aqa-Ib** [comment]
Alright I got it, it works when launched with a relative path to the current working directory:

- `cd /home/user/game/`
- `ulwgl-run /x64/launcher.exe`

_Notice that the ulwgl-run command can't contain the full path in this case or it won't work._

**Comment by Aqa-Ib** [comment]
It would be nice if the working directory could be defined into the ulwgl-run command though, like Steam and Lutris do.

**Comment by R1kaB3rN** [comment]
What's your specific use case? Personally, I've never used it when launching my games so I never saw the point of changing to a specific directory before launching a game.

I can give you this feature if it's not too niche, but it'd be via the configuration file usage. For example:

> [ulwgl]
> exe = "/home/user/game/x64/launcher.exe"  
> start = "/home/user/game/" # Directory to change to
> proton = "/home/user/.local/share/Steam/compatibilitytools.d/GE-Proton9-1"
> prefix = "/home/user/.wine"
> game_id = "0"

**Comment by Aqa-Ib** [comment]
For some games I need to give ulwgl-run the .exe path relative to the cwd (current working directory).

When the game has its executable in a directory below the main game directory, it won't work without a relative path to the cwd. 

However, I have also found a bug with ulwgl-run that is not present in vanilla wine. Consider a game directory with the main executable called frontend.exe, and a secondary executable called start.exe.

- `cd /home/user/game`
- If I execute `ulwgl-run frontend.exe` it works.
- If I execute `ulwgl-run /home/user/game/frontend.exe` it will launch `start.exe` instead.

I guess the bug is caused because of a Proton specific feature that I am not aware of.

P.S.: I don't need the feature personally, I am fine giving it relative paths. However, I appreciate it. Thanks.

**Comment by loathingKernel** [comment]
> I can give you this feature if it's not too niche, but it'd be via the configuration file usage. For example:

FWIW, this is unusual but not niche in any way. Even recent triple-A games sometimes require such handling. IMO ULWGL should explore avenues to facilitate this for compatibility reasons.

I do not know how steam handles this though, if we are to emulate it, it is possible that the client itself switches to the working directory before executing the proton command.

**Comment by Aqa-Ib** [comment]
> I do not know how steam handles this though, if we are to emulate it, it is possible that the client itself switches to the working directory before executing the proton command.

It has to be something else, because in Steam you can pass the absolute path to the executable contained in a subfolder and it works. In wine and ulwgl you have to pass the relative path. In Lutris you can also pass the absolute path. Maybe we can ask @strycore about this.

**Comment by loathingKernel** [comment]
> It has to be something else, because in Steam you can pass the absolute path to the executable contained in a subfolder and it works.

Indeed  Steam changes into the working directory of the game, as evident by printing the environment when a game is launched through Proton. `PWD` env variable is set to the game's directory.

to test set the following command as the launch options for any Steam game `env | tee ~/game-env.txt #%command%`

**Comment by Aqa-Ib** [comment]
Maybe after changing to the working directory of the game, it does something like this:

Compare the two strings:
- "StartIn" = /home/user/game
- "Target" = /home/user/game/x64/launcher.exe

If $Target contains $StartIn, then remove the contents of $StartIn from $Target, leaving it as a relative path:
- "Target" = /x64/launcher.exe

Now we can just execute $Target relative to $StartIn, and it will work.

**Comment by loathingKernel** [comment]
The issue lies with the windows executable and how that hardcodes and resolves paths. That's the core of the issue. On the Linux side it doesn't matter you use an absolute path for the executable as long as you are in the correct working directory before running the executable.

**Comment by Aqa-Ib** [comment]
> On the Linux side it doesn't matter if you use an absolute path for the executable as long as you are in the correct working directory before calling the executable, because ultimately the executable will be called from the directory it expects.

I have tried it, and in a case like the one I mentioned, it only works using a relative path to the cwd. I don't know why, but it is the reality on my system.

**Comment by loathingKernel** [comment]
I re-read the issue and noticed that you tried that. So I deleted my reply since there is obviously some interaction that I am missing here. This might be related to anything, needs more investigation.

**Comment by Yoghurt4C** [comment]
you don't have to imagine this structure if you have Grim Dawn in your steam/gog library; it keeps *all* of the x64 binaries - including dependencies and apis - in a subdirectory, and you will not be able to get to the main menu using the x64 executable if you don't step the working directory out of the subdirectory with said executable. steam (not sure if it's a thing on gog) handles switching between x32 and x64 through a launch argument but it's rather dubious either way

**Comment by ltsdw** [comment]
As discussed with loathingKernel on discord, I think this is related to my use case, one addon of the game Elder Scrolls Online, TamrielTradeCenter, has a client that should be run, if I run it before the game with either runinprefix or run, and after that I launcher the game, it launches normally, but when I do the opposite it doesn't:

```
Application could not be started, or no application associated with the specified file.
ShellExecuteEx failed: File not found. 
```

Here is an example:

If I run the game first:

```
WINEPREFIX=/home/med_user/.local/share/Games/TESO/pfx GAMEID=umu-306130 PROTONPATH=/home/med_user/.local/opt/GE-Proton9-2 /home/med_user/.local/opt/umu/bin/./umu-run runinprefix "/home/med_user/.local/share/Games/TESO/pfx/drive_c/Program Files (x86)/Zenimax Online/The Elder Scrolls Online/game/client/eso64.exe"
```

then the Client.exe which is in another partition:

```
WINEPREFIX=/home/med_user/.local/share/Games/TESO/pfx GAMEID=umu-306130 PROTONPATH=/home/med_user/.local/opt/GE-Proton9-2 /home/med_user/.local/opt/umu/bin/./umu-run runinprefix "/storages/xfs_storage_partition/med_user/Desktop/Documents/Elder Scrolls Online/live/AddOns/TamrielTradeCentre/Client/Client.exe"   

...

Application could not be started, or no application associated with the specified file.
ShellExecuteEx failed: File not found. 
```

It won't work.

The inverse works though (running the Client.exe first and after that running the game):

```
WINEPREFIX=/home/med_user/.local/share/Games/TESO/pfx GAMEID=umu-306130 PROTONPATH=/home/med_user/.local/opt/GE-Proton9-2 /home/med_user/.local/opt/umu/bin/./umu-run runinprefix "/storages/xfs_storage_partition/med_user/Desktop/Documents/Elder Scrolls Online/live/AddOns/TamrielTradeCentre/Client/Client.exe"   
```

```
WINEPREFIX=/home/med_user/.local/share/Games/TESO/pfx GAMEID=umu-306130 PROTONPATH=/home/med_user/.local/opt/GE-Proton9-2 /home/med_user/.local/opt/umu/bin/./umu-run runinprefix "/home/med_user/.local/share/Games/TESO/pfx/drive_c/Program Files (x86)/Zenimax Online/The Elder Scrolls Online/game/client/eso64.exe"
```

Also as discussed there, if I want to run the Client.exe I can also do this, which will work doesn't matter what working directory I'm in:

```
WINEPREFIX=/home/med_user/.local/share/Games/TESO/pfx GAMEID=umu-306130 PROTONPATH=/home/med_user/.local/opt/GE-Proton9-2 /home/med_user/.local/opt/umu/bin/./umu-run runinprefix start.exe /d "C:\users\steamuser\Documents\Elder Scrolls Online\live\AddOns\TamrielTradeCentre\Client" "C:\users\steamuser\Documents\Elder Scrolls Online\live\AddOns\TamrielTradeCentre\Client\Client.exe"
```

**Comment by Aqa-Ib** [comment]
Can you explain what "runinprefix" does and why you call start.exe ?

**Comment by loathingKernel** [comment]
To understand what these verbs do better, look at proton's `__main__` function. In short `runinprefix` does exactly what it implies, it runs something in the game's prefix.

`start.exe` here is to use some of its arguments, for a list and a brief explanation, run `wine start.exe /h`

**Comment by Ketrel** [comment]
For what it's worth, you can do this still

`/path/to/umu-run start /b /d "C:/Some/Folder/Path" "C:/Some/Binary.exe"`

**Comment by loathingKernel** [comment]
The pressure-vessel docs also mention this constraint (first two points), linking it [here](https://gitlab.steamos.cloud/steamrt/steam-runtime-tools/-/blob/main/docs/pressure-vessel.md#design-constraints) for discoverability.

**Comment by R1kaB3rN** [comment]
The first point in the documentation can easily be implemented, but the current working directory in a subdirectory mess should be addressed in a protonfix unless there's a reliable way to handle it. For instance, does the Steam client reference a whitelist or have a configuration file for the correct directory to change to? For each game on Steam, there's a manifest file that contains an `installdir` field that we can try change to, but I'm unsure.

@Aqa-Ib for Grim Dawn, does the appmanifest_219990.acf file have an `installdir` field that references the correct directory to start the game?

**Comment by loathingKernel** [comment]
As far as other stores are concerned, they should provide the working directory in their metadata in one way or another. I know for example that legendary does have that information and sets that as the working directory before launching. It is safe to assume that other clients do that too, and that's why we haven't had a big influx of comments on this issue. This information is also required on windows, as the working directory is part of the desktop and menu shortcut specification. This is mostly relevant when using umu as a stand-alone application from the terminal.

Thus, a likely solution would be to allow users to arbitrarily set a working directory from the command line, for example.

**Comment by Aqa-Ib** [comment]
> @Aqa-Ib for Grim Dawn, does the appmanifest_219990.acf file have an installdir field that references the correct directory to start the game?

I own Grim Dawn from GOG, unfortunately, it does not have a manifest file.
Cheers!

**Comment by spiffeeroo** [comment]
SteamDB shows the installdir config info.

https://steamdb.info/app/219990/config/

> installdir 	Grim Dawn

**Comment by R1kaB3rN** [comment]
Hmm, I believe it may be possible solve this problem for Steam and GOG's case. Steam provides an [installscript.vdf](https://partner.steamgames.com/doc/sdk/installscripts) that contains metadata that is defined by the developer which may include its subdirectory. For example:

```
"InstallScript"
{
	"Registry"
	{
		"HKEY_LOCAL_MACHINE\\SOFTWARE\\Foo\\Bar"
		{
			"string"
			{
				"Install_Path"      "%INSTALLDIR%"
				"Exe_Path"          "%INSTALLDIR%\\foobar.exe"
			}

			"dword"
			{
				"PatchVersion"      "12"
			}
		}
	}

	"Run Process"
	{
		"DirectX"
		{
			"HasRunKey"     "HKEY_LOCAL_MACHINE\\Software\\Valve\\Steam\\Apps\\480"
			"Process 1"     "%INSTALLDIR%\\DirectX\\DXSETUP.exe"
			"Command 1"     "/silent"
			"NoCleanUp"     "1"
		}
	}
}
```
 As for GOG, they should include an `.info` file in the game's top-level directory:

```
{
    "buildId": "55626781736221498",
    "clientId": "55466758769298188",
    "gameId": "1697970811",
    "language": "English",
    "languages": [
        "en-US"
    ],
    "name": "Flowers - Le Volume Sur Printemps",
    "playTasks": [
        {
            "category": "game",
            "isPrimary": true,
            "languages": [
                "zh-Hans",
                "en-US"
            ],
            "name": "Flowers - Le Volume Sur Printemps",
            "path": "FLOWERS.exe",
            "type": "FileTask"
        }
    ],
    "rootGameId": "1697970811",
    "version": 1
}
```

@Aqa-Ib What's the value of `path` for Grim Dawn?

**Comment by R1kaB3rN** [comment]
[Steamdb even shows a launch option that's the relative execution path](https://steamdb.info/app/219990/config/) of the 64-bit executable -- `x64/grim dawn.exe`. It's possible that path may be in Steam's `installscript.vdf` too.

**Comment by Aqa-Ib** [comment]
Correct. The file is named `goggame-1449651388.info` and the path value inside is `"path": "x64/Grim Dawn.exe"`. There is also the 32 bit executable path (the same as in steamdb).

**Comment by R1kaB3rN** [comment]
@Aqa-Ib Just made a PR to address the problem for GOG games. If you can make time to test it, that'd' be great. Parsing Steam's .vdf files will need to be addressed later -- I have to find some real examples.

Also, do you mind sharing the contents of the *.info file for Grim Dawn? GOG doesn't seem to make those files easily available, but I believe there should be a `workingDir` field.

**Comment by R1kaB3rN** [comment]
Actually, for Steam's case, I believe umu would need to get the [subdirectory information via a game's launch options](https://steamdb.info/app/1245620/config/) but unfortunately it doesn't appear it will be simple to retrieve.

**Comment by R1kaB3rN** [comment]
Closing as won’t fix and this will be a constraint with this tool. For this to work, umu-launcher, when used standalone, needs to be provided the information from a reliable source but only clients like Heroic Games Launcher are in a position to do that. Therefore, users who are using this tool standalone are advised to execute within the base game directory by default or execute the game’s lnk file using start.exe, if there’s any. Otherwise, if the game doesn’t launch, users must research the correct directory themselves or consider a client such as HGL that can automate this step, _if_ you followed the appropriate steps to install the game that is.

**Comment by tryashtar** [comment]
In my use case, I want to use umu standalone to launch a game in a completely separate working directory. With vanilla wine, this works just fine. With umu, it seems to ignore the provided working directory. The `start` shortcut workaround doesn't work either. Umu can't find the target of the shortcut, while again vanilla wine can just fine. I think it's related to sandboxing.

If these third-party clients can set the working directory in a custom way, how can I?

**Comment by Ketrel** [comment]
How are you using `start.exe` to do that?

Here's how my launch command for The Binding of Isaac looks  (I'm not including envars I set for prefix and whatnot)

`/usr/bin/umu-run start /b /d "C:/GOG Games/The Binding of Isaac Rebirth" "isaac-ng_launcher.exe"`

(note the use of the /d switch to set the directory)

**Comment by tryashtar** [comment]
I tried
```sh
cd '/d/Games/Kingdom Hearts/Launch/1.5+2.5'
umu-run start /b /d 'Z:\d\Games\Kingdom Hearts\Launch\1.5+2.5' 'Z:\d\Games\Kingdom Hearts\Installs\1.5+2.5\KINGDOM HEARTS FINAL MIX.exe'
```
Output says:
```
Application could not be started, or no application associated with the specified file.
ShellExecuteEx failed: Path not found.
```
Notice that I want the working directory in `Launch`, but the exe is located in `Installs`. When I replace `umu-run` with `wine`, this works correctly.

**Comment by Ketrel** [comment]
Two thing:

1. If you use `umu-run` to start `winecfg` are your drive mappings correct there?
2. Are you manually setting the prefix with an envar?

**Comment by tryashtar** [comment]
1. That shows `Z:` as `/` which seems right
2. Yeah, just like yours I elided the `WINEPREFIX='/d/Games/Kingdom Hearts/Prefix'` and other vars before the command

**Comment by Ketrel** [comment]
That looks wrong.

Why are you using the Z: mapping?

Why doesn't your path to the executable have the `drive_c` directory as part of it?

**Comment by tryashtar** [comment]
I'm using the `Z:` mapping because my games are installed to my linux partition, they aren't in the wineprefix. Again, this all works correctly with vanilla `wine`.

**Comment by Ketrel** [comment]
Can you paste a screenshot of the drive mappings

**Comment by tryashtar** [comment]
```sh
WINEPREFIX='/d/Games/Kingdom Hearts/Prefix' umu-run winecfg
```

<img width="163" height="112" alt="Image" src="https://github.com/user-attachments/assets/0b5d9b8f-b476-4024-854d-2c96c8ff902c" />

**Comment by Ketrel** [comment]
I'm curious, if you call up a cmd prompt, can you navigate to the directory you're trying to hit via the z mapping?  If not, that might make it more clear at what point something is going wrong.

**Comment by tryashtar** [comment]
Yes, I do that explicitly right before running the command. And again, normal wine runs the game with that working directory just fine as expected.

**Comment by Ketrel** [comment]
No I mean do `umu-run start /b cmd.exe`

Then from the command prompt do

`z:`

then try to navigate each step of the way, see where it fails.

**Comment by tryashtar** [comment]
Oh, gotcha. And oops, I accidentally deleted one of my other replies.

It starts in that directory, and I can navigate there manually one step at a time too.

<img width="641" height="206" alt="Image" src="https://github.com/user-attachments/assets/3a7fa894-d2ed-4608-a19b-39ff5b5aab49" />

I am unable to navigate to `Installs`, where the exe lives. Something about umu is sandboxing the wineprefix to only expose the working directory.

<img width="541" height="53" alt="Image" src="https://github.com/user-attachments/assets/46e4b590-7f70-49b4-9dbe-b3294c1757d4" />

**Comment by Ketrel** [comment]
do a `dir` in the Kingdom Hearts directory.

**Comment by tryashtar** [comment]
```sh
cd '/d/Games/Kingdom Hearts/Launch/1.5+2.5'
WINEPREFIX='/d/Games/Kingdom Hearts/Prefix' umu-run start cmd.exe
# in cmd
cd ..
cd ..
dir
```

<img width="429" height="129" alt="Image" src="https://github.com/user-attachments/assets/bdd1f0b5-dda0-445f-9539-4f29f1872f99" />

```sh
cd '/d/Games/Kingdom Hearts/Launch/1.5+2.5'
WINEPREFIX='/d/Games/Kingdom Hearts/Prefix' wine start cmd.exe
# in cmd
cd ..
cd ..
dir
```

<img width="668" height="332" alt="Image" src="https://github.com/user-attachments/assets/f1a95cd1-c76d-49bd-9b67-2284e27ac215" />

**Comment by Ketrel** [comment]
That's definitely odd. You can't CD into that Installs directory?  On the Linux end, the permissions are ok (especially execute)?

**Comment by tryashtar** [comment]
I can with wine, but not umu, as it can't even find it.

**Comment by Ketrel** [comment]
That's not what I asked....?

**Comment by tryashtar** [comment]
Yes, I can CD into it on the Linux side.

**Comment by Ketrel** [comment]
Is it a real directory or is it a symlink to somewhere else?

**Comment by tryashtar** [comment]
It's a normal real directory.

**Comment by Ketrel** [comment]
That's extremely strange.  Can you double check the permissions on that from the Linux terminal?
Does anything show up in dmesg when it fails to work?

**Comment by loathingKernel** [comment]
This is an issue tracker, not a forum for generic help. And if you had looked through other issues instead of spamming you would have found https://github.com/Open-Wine-Components/umu-launcher/issues/195 or https://github.com/Open-Wine-Components/umu-launcher/wiki/Frequently-asked-questions-(FAQ)#how-do-i-make-more-files-available-to-the-steam-runtime-container
